## **Ejercicio 1. Carga, comprensión e integración de los datos**

---

## **1.1. Carga de los archivos**

In [1]:
from pathlib import Path

import pandas as pd

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 160)

RAIZ = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()
DIR_RAW = RAIZ / "data" / "raw"

RUTA_VIDEOS = DIR_RAW / "youtube_videos.csv"
RUTA_COMENTARIOS = DIR_RAW / "youtube_comments.csv"

OPCIONES_LECTURA = dict(dtype=str, keep_default_na=False, encoding="utf-8-sig")

videos = pd.read_csv(RUTA_VIDEOS, **OPCIONES_LECTURA)
comentarios = pd.read_csv(RUTA_COMENTARIOS, **OPCIONES_LECTURA)

print(f"[cargado]  youtube_videos.csv    -> {videos.shape[0]:>4} filas, {videos.shape[1]:>2} columnas")
print(f"[cargado]  youtube_comments.csv  -> {comentarios.shape[0]:>4} filas, {comentarios.shape[1]:>2} columnas")

[cargado]  youtube_videos.csv    ->  293 filas, 20 columnas
[cargado]  youtube_comments.csv  ->  406 filas, 17 columnas


In [2]:
with open(RUTA_VIDEOS, "rb") as f:
    primeros_bytes = f.read(12)

print("Primeros bytes crudos :", primeros_bytes)
print("Trae BOM (EF BB BF)   :", primeros_bytes.startswith(b"\xef\xbb\xbf"))
print("Primera columna       :", repr(videos.columns[0]))

Primeros bytes crudos : b'\xef\xbb\xbfvideo_id,'
Trae BOM (EF BB BF)   : True
Primera columna       : 'video_id'


In [3]:
print("COLUMNAS DE youtube_videos.csv")
for i, col in enumerate(videos.columns, start=1):
    print(f"  {i:>2}. {col}")

print()
print("COLUMNAS DE youtube_comments.csv")
for i, col in enumerate(comentarios.columns, start=1):
    print(f"  {i:>2}. {col}")

COLUMNAS DE youtube_videos.csv
   1. video_id
   2. title
   3. channel_name
   4. channel_id
   5. source_query
   6. source_group
   7. dataset_sources
   8. channel_handle
   9. published_time
  10. view_count_text
  11. description_snippet
  12. video_url
  13. query_hits
  14. keywords
  15. description
  16. view_count
  17. publish_date
  18. upload_date
  19. category
  20. owner_handle

COLUMNAS DE youtube_comments.csv
   1. video_id
   2. comment_id
   3. video_title
   4. channel_name
   5. channel_id
   6. author_name
   7. author_channel_id
   8. text
   9. source_query
  10. source_group
  11. dataset_sources
  12. author_handle
  13. published_text
  14. like_count_text
  15. reply_count
  16. is_pinned
  17. viewer_rating


In [4]:
videos.head(3)

,video_id,title,channel_name,channel_id,source_query,source_group,dataset_sources,channel_handle,published_time,view_count_text,description_snippet,video_url,query_hits,keywords,description,view_count,publish_date,upload_date,category,owner_handle
0,-5puKGEqcUc,INSIVUMEH pronostica incremento de lluvias par...,T13 Noticias Guatemala,UCq0Cm-3SKthEySQc2JZBi1A,guatemala lluvias,topic,youtube_guatemala.csv | youtube_guatemala_lab....,/@T13NoticiasGuatemala,hace 2 días,"2,390 vistas",El Departamento de Pronóstico de INSIVUMEH pre...,https://www.youtube.com/watch?v=-5puKGEqcUc,"[""guatemala lluvias""]","[""Canícula prolongada"", ""Chapin tv"", ""Fenómeno...",El Departamento de Pronóstico de INSIVUMEH pre...,2357,2026-08-28T22:00:20-07:00,2026-08-28T22:00:20-07:00,News & Politics,/@T13NoticiasGuatemala
1,-E7OPOLjMug,BERNARDO ARÉVALO CALIFICA CAMBIO EN EL MP COMO...,IDocumenta,UCgItjn_ZWFWcv1MlpKIkqgQ,@GobiernodelaRepublicadeGuatema,topic,youtube_target_channels.csv,/@iDocumenta,hace 3 meses,4 vistas,"El presidente de Guatemala, Bernardo Arévalo, ...",https://www.youtube.com/watch?v=-E7OPOLjMug,"[""@GobiernodelaRepublicadeGuatema""]",[],"El presidente de Guatemala, Bernardo Arévalo, ...",4,2026-05-13T16:00:03-07:00,2026-05-13T16:00:03-07:00,People & Blogs,/@iDocumenta
2,-KDglrIzRKo,¡HISTÓRICO! Mexico recupera petróleo robado po...,México Poder,UC-DpoeBbCMOMOH1-j71KPqA,guatemala noticias,topic,youtube_guatemala.csv | youtube_guatemala_lab....,/@M%C3%A9xicoPoder,hace 1 día,"29,736 vistas",México #PetróleoMexicano #SoberaníaEnergética ...,https://www.youtube.com/watch?v=-KDglrIzRKo,"[""guatemala noticias""]","[""Mexico recupera petroleo"", ""petroleo robado ...",#México #PetróleoMexicano #SoberaníaEnergética...,29736,2026-08-29T17:00:38-07:00,2026-08-29T17:00:38-07:00,People & Blogs,/@M%C3%A9xicoPoder


In [5]:
comentarios.head(3)

,video_id,comment_id,video_title,channel_name,channel_id,author_name,author_channel_id,text,source_query,source_group,dataset_sources,author_handle,published_text,like_count_text,reply_count,is_pinned,viewer_rating
0,j43HgwYFKfk,Ugw-J65a1iYL9hqhELh4AaABAg,La cooptación de Walter Mazariegos en la USAC,Quorum,UCE4rsXcgDb6e1-a9iTbWzfg,@MarcosCarillo-b1r,UCdFlugHJJa4l3YqWuNRmvXw,Ese corrupto amigo de la vieja fiscal los teng...,@quorumgt,topic,j43HgwYFKfk_out_comments.csv | quorum_complete...,/@MarcosCarillo-b1r,hace 6 meses,,0,False,
1,06mFNPU0aB8,Ugw-ZT9t9wU2V-tCaUZ4AaABAg,Capturan a presuntos delincuentes disfrazados ...,Noti7,UCVpSRoZgngfSL03Nlbjtq9A,@RaulPerez-cw2vi,UCvl1tzQeBeGy6efPTRJXSCw,"Están jóvenes porque no buscan un trabajo, tu...",guatemala noticias,topic,youtube_guatemala_comments.csv | youtube_guate...,/@RaulPerez-cw2vi,hace 2 semanas,,0,False,
2,j43HgwYFKfk,Ugw0xaOb2CYXXoudtwJ4AaABAg,La cooptación de Walter Mazariegos en la USAC,Quorum,UCE4rsXcgDb6e1-a9iTbWzfg,@iamjimalesssa,UCRAquv8el-tQ30bN7MlmySQ,Me dejaron con ganas de demandar la ilegalidad...,@quorumgt,topic,j43HgwYFKfk_out_comments.csv | quorum_complete...,/@iamjimalesssa,hace 1 año,4,0,False,


`youtube_videos.csv` trae 293 filas y 20 variables, donde cada observación representa un video de
YouTube con sus atributos de publicación, sus métricas de consumo y la información de cómo fue
recolectado. `youtube_comments.csv` trae 406 filas y 17 variables, y en este caso cada observación
representa un comentario publicado en alguno de esos videos, aunque no todos son comentarios
principales, ya que 36 de los 406 son respuestas a otro comentario.

---

## **1.2. Unidad de observación, llave primaria y variables relevantes**

In [6]:
def diagnostico_llave(df, nombre_df, columna, rol):
    """Verifica unicidad y ausencia de vacios en una llave candidata."""
    serie = df[columna]
    vacios = int((serie.str.strip() == "").sum())
    duplicados = int(serie.duplicated().sum())

    print(f"{nombre_df}.{columna}  ({rol})")
    print(f"    filas totales     : {len(serie)}")
    print(f"    valores distintos : {serie.nunique()}")
    print(f"    duplicados        : {duplicados}")
    print(f"    vacios            : {vacios}")
    print(f"    sirve como llave  : {'SI' if duplicados == 0 and vacios == 0 else 'NO'}")
    print()


diagnostico_llave(videos, "videos", "video_id", "llave primaria")
diagnostico_llave(comentarios, "comentarios", "comment_id", "llave primaria")
diagnostico_llave(comentarios, "comentarios", "video_id", "llave foranea -> videos.video_id")

videos.video_id  (llave primaria)
    filas totales     : 293
    valores distintos : 293
    duplicados        : 0
    vacios            : 0
    sirve como llave  : SI

comentarios.comment_id  (llave primaria)
    filas totales     : 406
    valores distintos : 406
    duplicados        : 0
    vacios            : 0
    sirve como llave  : SI

comentarios.video_id  (llave foranea -> videos.video_id)
    filas totales     : 406
    valores distintos : 19
    duplicados        : 387
    vacios            : 0
    sirve como llave  : NO



In [7]:
def inventario_columnas(df):
    """Resume cada columna: valores distintos, vacios y si es constante."""
    filas = []
    for col in df.columns:
        serie = df[col]
        vacios = int((serie.str.strip() == "").sum())
        distintos = int(serie.nunique())
        filas.append({
            "columna": col,
            "distintos": distintos,
            "vacios": vacios,
            "pct_vacios": round(100 * vacios / len(serie), 1),
            "constante": distintos <= 1,
            "ejemplo": serie.iloc[0][:45],
        })
    return pd.DataFrame(filas)


inventario_columnas(videos)

,columna,distintos,vacios,pct_vacios,constante,ejemplo
0,video_id,293,0,0.0,False,-5puKGEqcUc
1,title,274,0,0.0,False,INSIVUMEH pronostica incremento de lluvias pa
2,channel_name,97,0,0.0,False,T13 Noticias Guatemala
3,channel_id,97,0,0.0,False,UCq0Cm-3SKthEySQc2JZBi1A
4,source_query,21,0,0.0,False,guatemala lluvias
5,source_group,3,0,0.0,False,topic
6,dataset_sources,23,0,0.0,False,youtube_guatemala.csv | youtube_guatemala_lab
7,channel_handle,97,0,0.0,False,/@T13NoticiasGuatemala
8,published_time,81,13,4.4,False,hace 2 días
9,view_count_text,260,13,4.4,False,"2,390 vistas"


In [8]:
inventario_columnas(comentarios)

,columna,distintos,vacios,pct_vacios,constante,ejemplo
0,video_id,19,0,0.0,False,j43HgwYFKfk
1,comment_id,406,0,0.0,False,Ugw-J65a1iYL9hqhELh4AaABAg
2,video_title,19,0,0.0,False,La cooptación de Walter Mazariegos en la USAC
3,channel_name,8,0,0.0,False,Quorum
4,channel_id,8,0,0.0,False,UCE4rsXcgDb6e1-a9iTbWzfg
5,author_name,332,0,0.0,False,@MarcosCarillo-b1r
6,author_channel_id,332,0,0.0,False,UCdFlugHJJa4l3YqWuNRmvXw
7,text,404,0,0.0,False,Ese corrupto amigo de la vieja fiscal los ten
8,source_query,6,0,0.0,False,@quorumgt
9,source_group,2,0,0.0,False,topic


In [9]:
print("REDUNDANCIAS EN videos")
print(f"  upload_date  == publish_date    : {(videos.upload_date == videos.publish_date).mean():.1%}")
print(f"  owner_handle == channel_handle  : {(videos.owner_handle == videos.channel_handle).mean():.1%}")
print()

print("CONSTANTES O VACIAS EN comentarios")
print(f"  is_pinned     -> valores distintos : {comentarios.is_pinned.unique().tolist()}")
print(f"  viewer_rating -> vacia en          : {(comentarios.viewer_rating.str.strip() == '').mean():.1%}")
print()

print("CONTEOS Y FECHAS ALMACENADOS COMO TEXTO")
print(f"  videos.view_count_text      : {videos.view_count_text.iloc[0]!r}")
print(f"  videos.published_time       : {videos.published_time.iloc[0]!r}")
print(f"  comentarios.like_count_text : {comentarios.like_count_text.head(4).tolist()}")
print(f"  comentarios.published_text  : {comentarios.published_text.iloc[0]!r}")
print()

print("NOMBRES VISIBLES QUE SON HANDLES")
print(f"  author_name que empiezan con '@' : {comentarios.author_name.str.startswith('@').mean():.1%}")

REDUNDANCIAS EN videos
  upload_date  == publish_date    : 100.0%
  owner_handle == channel_handle  : 100.0%

CONSTANTES O VACIAS EN comentarios
  is_pinned     -> valores distintos : ['False']
  viewer_rating -> vacia en          : 100.0%

CONTEOS Y FECHAS ALMACENADOS COMO TEXTO
  videos.view_count_text      : '2,390 vistas'
  videos.published_time       : 'hace 2 días'
  comentarios.like_count_text : [' ', ' ', '4', ' ']
  comentarios.published_text  : 'hace 6 meses'

NOMBRES VISIBLES QUE SON HANDLES
  author_name que empiezan con '@' : 100.0%


In [10]:
# El comment_id de una respuesta tiene la forma <id_del_padre>.<id_de_la_respuesta>
es_respuesta = comentarios.comment_id.str.contains(r"\.", regex=True)
id_padre = comentarios.loc[es_respuesta, "comment_id"].str.split(".").str[0]

print("ESTRUCTURA DE comment_id")
print(f"  largos observados                       : {comentarios.comment_id.str.len().value_counts().sort_index().to_dict()}")
print(f"  comentarios principales                 : {int((~es_respuesta).sum())}")
print(f"  respuestas                              : {int(es_respuesta.sum())}")
print(f"  respuestas cuyo padre esta en el archivo: {int(id_padre.isin(comentarios.comment_id).sum())} de {int(es_respuesta.sum())}")

ESTRUCTURA DE comment_id
  largos observados                       : {26: 370, 49: 36}
  comentarios principales                 : 370
  respuestas                              : 36
  respuestas cuyo padre esta en el archivo: 36 de 36


**Unidad de observación**: en `youtube_videos.csv` es un video de YouTube, y en
`youtube_comments.csv` es un comentario publicado en un video. Cabe mencionar que el enunciado
describe este segundo archivo como comentarios principales, pero al revisar el formato de
`comment_id` se observa que eso no se cumple del todo: 370 registros son comentarios de primer
nivel y 36 son respuestas, ya que su identificador toma la forma
`id_del_padre.id_de_la_respuesta` y en los 36 casos ese padre está presente en el mismo archivo.

**Llave primaria**: `video_id` en videos y `comment_id` en comentarios, ambas únicas y sin vacíos en
los 293 y 406 registros respectivamente. Adicional, `video_id` funciona como llave foránea dentro de
comentarios, donde toma 19 valores distintos.

**Variables en `youtube_videos.csv`**

- `video_id`: identificador único del video; es la llave primaria del archivo.
- `title`: título del video tal como se publicó.
- `channel_name`: nombre visible del canal que publicó el video.
- `channel_id`: identificador único del canal dueño del video.
- `source_query`: consulta o canal con el que se encontró el video durante la recolección.
- `source_group`: estrategia de búsqueda de la que proviene el video: `topic`, `official_gov` o `channel`.
- `dataset_sources`: archivos originales en los que aparecía el video antes de integrar las fuentes.
- `channel_handle`: handle del canal, con formato `/@nombre`.
- `published_time`: tiempo transcurrido desde la publicación, expresado de forma relativa.
- `view_count_text`: visualizaciones tal como las muestra YouTube, almacenadas como texto.
- `description_snippet`: fragmento abreviado de la descripción.
- `video_url`: dirección web completa del video.
- `query_hits`: lista de todas las consultas que recuperaron el video.
- `keywords`: lista de etiquetas asociadas al video; viene vacía en 162 de los 293 registros.
- `description`: descripción completa del video; viene vacía en 26 registros.
- `view_count`: visualizaciones como número entero.
- `publish_date`: fecha y hora de publicación en formato ISO 8601 con zona horaria.
- `upload_date`: fecha de carga; es idéntica a `publish_date` en el 100 % de los registros.
- `category`: categoría temática asignada por YouTube, de un total de 11 observadas.
- `owner_handle`: handle del propietario; es idéntico a `channel_handle` en el 100 % de los registros.

**Variables en `youtube_comments.csv`**

- `video_id`: identificador del video donde se publicó el comentario; es la llave foránea hacia videos.
- `comment_id`: identificador único del comentario; es la llave primaria del archivo y en las respuestas toma la forma `id_del_padre.id_de_la_respuesta`.
- `video_title`: título del video comentado; coincide con `title` en el 100 % de los registros.
- `channel_name`: nombre del canal dueño del video, no del autor del comentario.
- `channel_id`: identificador del canal dueño del video comentado.
- `author_name`: nombre visible del autor; en realidad es un handle, ya que empieza con `@` en el 100 % de los registros.
- `author_channel_id`: identificador único de la cuenta que escribió el comentario.
- `text`: contenido completo del comentario.
- `source_query`: consulta o canal con el que se obtuvieron los comentarios.
- `source_group`: tipo de fuente utilizada: `topic` o `channel`.
- `dataset_sources`: archivos de comentarios originales en los que aparecía el registro.
- `author_handle`: handle del autor, con formato `/@nombre`.
- `published_text`: tiempo transcurrido desde la publicación del comentario, expresado de forma relativa.
- `like_count_text`: "me gusta" del comentario como texto; 189 de los 406 vienen como un espacio en blanco.
- `reply_count`: número de respuestas que recibió el comentario; no identifica a sus autores, aunque 36 de esas respuestas sí aparecen como fila propia.
- `is_pinned`: si el comentario fue fijado por el canal; toma el valor `False` en los 406 registros.
- `viewer_rating`: valoración del usuario; está vacía en los 406 registros.

---

## **1.3. Relación entre canal, video, autor del comentario, comentario, categoría y consulta**

In [11]:
print("CARDINALIDADES OBSERVADAS")
print(f"  canales distintos                : {videos.channel_id.nunique()}")
print(f"  videos por canal (promedio)      : {len(videos) / videos.channel_id.nunique():.2f}")
print(f"  videos por canal (maximo)        : {videos.channel_id.value_counts().max()}")
print()
print(f"  autores distintos                : {comentarios.author_channel_id.nunique()}")
print(f"  comentarios por autor (promedio) : {len(comentarios) / comentarios.author_channel_id.nunique():.2f}")
print(f"  comentarios por autor (maximo)   : {comentarios.author_channel_id.value_counts().max()}")
print()
print(f"  categorias distintas             : {videos.category.nunique()}")
print(f"  consultas distintas              : {videos.source_query.nunique()}")

CARDINALIDADES OBSERVADAS
  canales distintos                : 97
  videos por canal (promedio)      : 3.02
  videos por canal (maximo)        : 32

  autores distintos                : 332
  comentarios por autor (promedio) : 1.22
  comentarios por autor (maximo)   : 6

  categorias distintas             : 11
  consultas distintas              : 21


In [12]:
canales_video = set(videos.channel_id)
canales_autor = set(comentarios.author_channel_id)

print("CANAL DUENO DEL VIDEO vs AUTOR DEL COMENTARIO")
print(f"  channel_id distintos en videos        : {len(canales_video)}")
print(f"  author_channel_id distintos           : {len(canales_autor)}")
print(f"  identificadores presentes en ambos    : {len(canales_video & canales_autor)}")

CANAL DUENO DEL VIDEO vs AUTOR DEL COMENTARIO
  channel_id distintos en videos        : 97
  author_channel_id distintos           : 332
  identificadores presentes en ambos    : 0


In [13]:
principales = comentarios[~es_respuesta]
respuestas = comentarios[es_respuesta]
autor_por_comentario = comentarios.set_index("comment_id").author_channel_id
autor_padre = respuestas.comment_id.str.split(".").str[0].map(autor_por_comentario)

print("RESPUESTAS DECLARADAS vs RESPUESTAS OBSERVABLES")
print(f"  comentarios con al menos una respuesta : {(principales.reply_count.astype(int) > 0).sum()} de {len(principales)}")
print(f"  respuestas declaradas por reply_count  : {principales.reply_count.astype(int).sum()}")
print(f"  respuestas presentes como fila         : {len(respuestas)}")
print(f"  respuestas ausentes del conjunto       : {principales.reply_count.astype(int).sum() - len(respuestas)}")
print(f"  pares autor -> autor recuperables      : {int((respuestas.author_channel_id.values != autor_padre.values).sum())}")

RESPUESTAS DECLARADAS vs RESPUESTAS OBSERVABLES
  comentarios con al menos una respuesta : 30 de 370
  respuestas declaradas por reply_count  : 51
  respuestas presentes como fila         : 36
  respuestas ausentes del conjunto       : 15
  pares autor -> autor recuperables      : 35


A grandes rasgos el esquema se lee así: cada canal publica videos, cada uno de esos videos recibe
comentarios, y cada comentario lo escribe un autor. La categoría y la consulta no son entidades sino
etiquetas que cuelgan del video, la primera describiendo su tema y la segunda el procedimiento con el
que se recolectó.

- **Canal -> video**, de uno a muchos. Los 293 videos provienen de 97 canales, con un promedio de 3.02
  videos por canal y un máximo de 32.
- **Video -> comentario**, de uno a muchos. Se materializa con `video_id` como llave foránea, y es la
  relación que se explota en el inciso 1.4.
- **Autor -> comentario**, de uno a muchos. Los 406 comentarios provienen de 332 autores, con un
  promedio de 1.22 comentarios por autor y un máximo de 6.
- **Video -> categoría**, de muchos a uno. Cada video tiene una sola de las 11 categorías observadas,
  por lo que la categoría no tiene identidad propia sino que funciona como etiqueta temática.
- **Consulta <-> video**, de muchos a muchos. Se observan 21 consultas distintas y `query_hits` guarda
  todas las que recuperaron cada video. Cabe mencionar que la consulta describe el muestreo y no el
  tema del video, ya que un video pudo entrar por `guatemala lluvias` sin que ese sea su tema central.
- **El canal dueño del video no es el autor del comentario**. `channel_id` y `author_channel_id` no
  comparten ningún identificador, de tal forma que los nodos de autor se deben construir
  exclusivamente con `author_channel_id`.
- **La relación entre autores es parcial y no se puede reconstruir por completo**. Los comentarios
  principales declaran 51 respuestas mediante `reply_count`, de las cuales 36 sí están presentes como
  fila y permiten recuperar 35 pares de autor hacia autor a partir del prefijo de `comment_id`. Las
  otras 15 no están en el conjunto, y `reply_count` por sí solo no dice quién respondió, es por esto
  que no puede leerse como una arista y la relación que los datos soportan de forma completa sigue
  siendo la que va de un autor hacia un video.

---

## **1.4. Integración de los conjuntos mediante `video_id`**

In [14]:
COLUMNAS_COMPARTIDAS = sorted(set(videos.columns) & set(comentarios.columns))
print("Columnas presentes en ambos archivos:", COLUMNAS_COMPARTIDAS)

Columnas presentes en ambos archivos: ['channel_id', 'channel_name', 'dataset_sources', 'source_group', 'source_query', 'video_id']


In [15]:
integrado = comentarios.merge(
    videos,
    on="video_id",
    how="left",
    suffixes=("_com", "_vid"),
    indicator=True,
)

emparejados = int((integrado["_merge"] == "both").sum())
sin_pareja = int((integrado["_merge"] == "left_only").sum())

print(f"[integrado]  {integrado.shape[0]} filas, {integrado.shape[1]} columnas")
print()
print("RESULTADO DE LA INTEGRACION")
print(f"  comentarios en el archivo original : {len(comentarios)}")
print(f"  comentarios asociados a un video   : {emparejados}  ({emparejados / len(comentarios):.2%})")
print(f"  comentarios huerfanos              : {sin_pareja}")

[integrado]  406 filas, 37 columnas

RESULTADO DE LA INTEGRACION
  comentarios en el archivo original : 406
  comentarios asociados a un video   : 406  (100.00%)
  comentarios huerfanos              : 0


In [16]:
print("CONTROL DE INTEGRIDAD DE LA UNION")
print(f"  filas antes del merge     : {len(comentarios)}")
print(f"  filas despues del merge   : {len(integrado)}")
print(f"  sin duplicacion de filas  : {'SI' if len(integrado) == len(comentarios) else 'NO'}")
print(f"  comment_id sigue unico    : {'SI' if integrado.comment_id.is_unique else 'NO'}")
print(f"  columnas con algun nulo   : {int(integrado.isna().any().sum())}")

CONTROL DE INTEGRIDAD DE LA UNION
  filas antes del merge     : 406
  filas despues del merge   : 406
  sin duplicacion de filas  : SI
  comment_id sigue unico    : SI
  columnas con algun nulo   : 0


In [17]:
videos_con_comentarios = comentarios.video_id.nunique()
videos_totales = videos.video_id.nunique()

print("COBERTURA DEL CATALOGO DE VIDEOS")
print(f"  videos en el catalogo              : {videos_totales}")
print(f"  videos con al menos un comentario  : {videos_con_comentarios}  ({videos_con_comentarios / videos_totales:.2%})")
print(f"  videos sin ningun comentario       : {videos_totales - videos_con_comentarios}  ({1 - videos_con_comentarios / videos_totales:.2%})")

COBERTURA DEL CATALOGO DE VIDEOS
  videos en el catalogo              : 293
  videos con al menos un comentario  : 19  (6.48%)
  videos sin ningun comentario       : 274  (93.52%)


In [18]:
print("COINCIDENCIA ENTRE COLUMNAS HOMONIMAS TRAS EL MERGE")
print(f"  video_title == title                : {(integrado.video_title == integrado.title).mean():.1%}")
print(f"  channel_name_com == channel_name_vid : {(integrado.channel_name_com == integrado.channel_name_vid).mean():.1%}")
print(f"  channel_id_com   == channel_id_vid   : {(integrado.channel_id_com == integrado.channel_id_vid).mean():.1%}")
print(f"  source_query_com == source_query_vid : {(integrado.source_query_com == integrado.source_query_vid).mean():.1%}")
print(f"  source_group_com == source_group_vid : {(integrado.source_group_com == integrado.source_group_vid).mean():.1%}")
print(f"  dataset_sources_com == ..._vid       : {(integrado.dataset_sources_com == integrado.dataset_sources_vid).mean():.1%}")

COINCIDENCIA ENTRE COLUMNAS HOMONIMAS TRAS EL MERGE
  video_title == title                : 100.0%
  channel_name_com == channel_name_vid : 100.0%
  channel_id_com   == channel_id_vid   : 100.0%
  source_query_com == source_query_vid : 53.7%
  source_group_com == source_group_vid : 53.7%
  dataset_sources_com == ..._vid       : 0.0%


In [19]:
conteo_por_video = comentarios.video_id.value_counts()
autores_por_video = comentarios.groupby("video_id").author_channel_id.nunique()

resumen_videos_comentados = (
    videos[videos.video_id.isin(conteo_por_video.index)]
    .assign(
        n_comentarios=lambda d: d.video_id.map(conteo_por_video),
        n_autores=lambda d: d.video_id.map(autores_por_video),
    )
    .sort_values("n_comentarios", ascending=False)
    .loc[:, ["video_id", "channel_name", "category", "n_comentarios", "n_autores", "title"]]
    .reset_index(drop=True)
)

resumen_videos_comentados["pct_del_total"] = (
    100 * resumen_videos_comentados.n_comentarios / len(comentarios)
).round(1)

resumen_videos_comentados

,video_id,channel_name,category,n_comentarios,n_autores,title,pct_del_total
0,n8iP75gIpmw,Quorum,News & Politics,161,128,Qué rico come tu diputado,39.7
1,j43HgwYFKfk,Quorum,News & Politics,50,49,La cooptación de Walter Mazariegos en la USAC,12.3
2,6W4u8sGEnGM,Gobierno de la República de Guatemala,Entertainment,45,32,Inician los trabajos de recuperación del Puent...,11.1
3,OkXlHx0hx-8,Noticias Telemundo,News & Politics,25,18,EE.UU. envía a mexicanos deportados a Guatemal...,6.2
4,PjmxCj-a9Hg,Gobierno de la República de Guatemala,Entertainment,25,19,Conferencia de Prensa del Gobierno de Guatemal...,6.2
5,lj983NWyAQY,Municipalidad de Guatemala,News & Politics,25,25,Plan 2032 Ciudad de Guatemala,6.2
6,yLZS3JiEBg8,Quorum,News & Politics,16,16,Arroz con pollo a la MONOPOLIO,3.9
7,06mFNPU0aB8,Noti7,News & Politics,14,13,Capturan a presuntos delincuentes disfrazados ...,3.4
8,ndAZjHqzzT8,Quorum,News & Politics,12,10,Internet: escoger el menos malo,3.0
9,b_Hkrkl3adA,PrensaLibreOficial,News & Politics,7,7,Bloqueos en Guatemala este 31 de agosto por al...,1.7


In [20]:
print("CONCENTRACION DE LA PARTICIPACION")
print(f"  comentarios por video (promedio) : {conteo_por_video.mean():.2f}")
print(f"  comentarios por video (mediana)  : {conteo_por_video.median():.0f}")
print(f"  comentarios por video (maximo)   : {conteo_por_video.max()}")
print(f"  concentra el video mas comentado : {conteo_por_video.max() / len(comentarios):.2%}")
print(f"  concentran los 3 mas comentados  : {conteo_por_video.head(3).sum() / len(comentarios):.2%}")
print(f"  canales entre los videos comentados : {videos[videos.video_id.isin(conteo_por_video.index)].channel_name.nunique()}")
print()
print("Comentarios por canal dueño del video:")
print(integrado.channel_name_vid.value_counts().to_string())

CONCENTRACION DE LA PARTICIPACION
  comentarios por video (promedio) : 21.37
  comentarios por video (mediana)  : 7
  comentarios por video (maximo)   : 161
  concentra el video mas comentado : 39.66%
  concentran los 3 mas comentados  : 63.05%
  canales entre los videos comentados : 8

Comentarios por canal dueño del video:
channel_name_vid
Quorum                                   256
Gobierno de la República de Guatemala     70
Noticias Telemundo                        25
Municipalidad de Guatemala                25
Noti7                                     14
PrensaLibreOficial                         7
TN23 Guatemala                             7
Noticiero Guatevisión 13 Hrs.              2


La integración se hizo con un `merge` por `video_id` tomando comentarios como tabla izquierda, ya que
lo que se busca comprobar es cuántos comentarios lograron asociarse a un video y de esta forma
ninguno se pierde en la operación. **Los 406 comentarios se asociaron a un video, es decir, el
100 %, sin ningún huérfano y sin duplicar filas.**

No obstante, en la dirección contraria la cobertura es distinta: solo 19 de los 293 videos recibieron
al menos un comentario, un 6.48 % del catálogo. El reparto entre esos 19 videos es el siguiente.

- 161 comentarios: *Qué rico come tu diputado* (Quorum)
- 50 comentarios: *La cooptación de Walter Mazariegos en la USAC* (Quorum)
- 45 comentarios: *Inician los trabajos de recuperación del Puente Belice II.* (Gobierno de la República de Guatemala)
- 25 comentarios: *EE.UU. envía a mexicanos deportados a Guatemala antes de su regreso a México* (Noticias Telemundo)
- 25 comentarios: *Conferencia de Prensa del Gobierno de Guatemala. #LaRondaGt* (Gobierno de la República de Guatemala)
- 25 comentarios: *Plan 2032 Ciudad de Guatemala* (Municipalidad de Guatemala)
- 16 comentarios: *Arroz con pollo a la MONOPOLIO* (Quorum)
- 14 comentarios: *Capturan a presuntos delincuentes disfrazados de mujer señalados de cometer asalto* (Noti7)
- 12 comentarios: *Internet: escoger el menos malo* (Quorum)
- 7 comentarios: *Bloqueos en Guatemala este 31 de agosto por alza en combustibles afectan rutas principales* (PrensaLibreOficial)
- 7 comentarios: *Capturan a ladrón que había quedado grabado mientras robaba en una parroquia de Retalhuleu* (TN23 Guatemala)
- 6 comentarios: *Caminar en una ciudad hecha para carros* (Quorum)
- 4 comentarios: *I'x K'at: el primer equipo guatemalteco de pelota maya, conformado únicamente por mujeres.* (Quorum)
- 3 comentarios: *10 Preguntas a un año del Paro Nacional* (Quorum)
- 2 comentarios: *Noticiero en Directo 1 pm, 28 de Agosto de 2026* (Noticiero Guatevisión 13 Hrs.)
- 1 comentario: *¿Quiénes pagan más en Centroamérica?* (Quorum)
- 1 comentario: *Cruzando la ciudad a puro Transmetro* (Quorum)
- 1 comentario: *SHAI WA: la vecina queer de Casa Presidencial* (Quorum)
- 1 comentario: *Edén por Salud: empleo inclusivo para personas con discapacidad en Antigua* (Quorum)

Esos 19 videos pertenecen a 8 canales, el más comentado concentra el 39.66 % de los comentarios y los
tres primeros juntos un 63.05 %.